### PR TITLE
@l2succes => [Messaging] Updating messaging connection + attachments

### PR DESCRIPTION
**Pull request by mzikherman**
This depends on https://github.com/artsy/impulse/pull/262 (already merged) which adds getting a fuller message details response.

Also depends on https://github.com/artsy/impulse/pull/265 to get some additional fields in there.

Once ^ is merged + deployed to staging, I can get my Emission in sync and open that PR, and then this one (and Emission) will be good to review/sync together.

But, this probably won't change too much (and I do have Emission sorta working currently), so might as well get it out there.

### CODE REVIEW

File: schema/me/conversation.js [hunk 1]
```diff
@@ -37,25 +37,54 @@ export const BuyerOutcomeTypes = new GraphQLEnumType({
...
 
+export const AttachmentType = new GraphQLObjectType({
+  name: "AttachmentType",
+  desciption: "Fields of an attachment (currently from Radiation)",
+  fields: {
+    id: {
+      description: "Attachment id.",
+      type: new GraphQLNonNull(GraphQLString),
+    },
+    content_type: {
```
Comment (mzikherman) [inline schema/me/conversation.js]: We'll be able to use `content_type` and `download_url` to offer PDF/Image previews.

File: schema/me/conversation.js [hunk 2]
```diff
@@ -37,25 +37,54 @@ export const BuyerOutcomeTypes = new GraphQLEnumType({
...
-      type: new GraphQLNonNull(GraphQLString),
+    is_from_user: {
+      description: "True if message is from the user to the partner.",
+      type: GraphQLBoolean,
+      resolve: ({ from_email_address, conversation_from_address }) => {
+        from_email_address === conversation_from_address
+      },
     },
-    radiation_message_id: {
+    raw_text: {
```
Comment (mzikherman) [inline schema/me/conversation.js]: This replaces the `snippet` and is where we'll be doing sanitization, presumably.

File: schema/me/conversation.js [hunk 3]
```diff
@@ -173,12 +202,24 @@ export const ConversationFields = {
...
     args: pageable(),
-    resolve: ({ messages }, options) => {
-      const impulseOptions = parseRelayOptions(options)
-      return connectionFromArraySlice(messages, options, {
-        arrayLength: messages.length,
-        sliceStart: impulseOptions.offset,
-      })
+    resolve: ({ id, from_email_address }, options, req, { rootValue: { accessToken } }) => {
+      const { page, size, offset } = parseRelayOptions(options)
+      const impulseParams = { page, size, conversation_id: id }
```
Comment (mzikherman) [inline schema/me/conversation.js]: We call Impulse with the `conversation_id`, plus pagination params. We use `parseRelayOptions` to take Relay cursor-style pagination and transform it to `page`, `size` and `offset`.

The API takes combinations of `page` and `size`, but we need `offset` a few lines below for the `sliceStart` Relay helper...../shrug

File: schema/me/conversation.js [hunk 4]
```diff
@@ -173,12 +202,24 @@ export const ConversationFields = {
...
+      return gravity
+        .with(accessToken, { method: "POST" })("me/token", {
+          client_application_id: IMPULSE_APPLICATION_ID,
+        })
+        .then(data => {
+          return impulse
+            .with(data.token, { method: "GET" })(`message_details`, impulseParams)
+            .then(({ total_count, message_details }) => {
```
Comment (mzikherman) [inline schema/me/conversation.js]: The response will have these two fields...pretty straightforward.

File: schema/me/conversation.js [hunk 5]
```diff
@@ -173,12 +202,24 @@ export const ConversationFields = {
...
+              message_details["conversation_from_address"] = from_email_address
```
Comment (mzikherman) [inline schema/me/conversation.js]: This deserves some explanation.

We inject, as `conversation_from_address`, into each message payload, the _unmasked_ email of the originator of the conversation. This is presumably the user's email address.

Thus, we can tell on a _per-message_ basis (in https://github.com/artsy/metaphysics/pull/680/files#diff-e1fb60e2b771055ea7d7b51ceec6aa31R78), if the message is from a user or not (and can render their name, or the partner logo).

Believe it or not this is the best way we can currently check the `from_type` of a particular message...we'll most likely add something more structured that will let us determine this soon.

File: schema/me/conversation.js (hunk 3 above)
Comment (l2succes) [inline schema/me/conversation.js]: 👍🏿 had to do the same of the conversations resolver as well, looks good

File: schema/me/conversation.js (hunk 5 above)
Comment (l2succes) [inline schema/me/conversation.js]: Got it, thanks for the explanation. So we just need this field to figure out if the message came from the user essentially?

File: schema/me/conversation.js [hunk 6]
```diff
@@ -37,25 +37,54 @@ export const BuyerOutcomeTypes = new GraphQLEnumType({
...
+      resolve: ({ from_email_address, conversation_from_address }) => {
+        from_email_address === conversation_from_address
+      },
     },
-    radiation_message_id: {
+    raw_text: {
+      description: "Full unsanitized text.",
       type: new GraphQLNonNull(GraphQLString),
     },
+    attachments: {
```
Comment (l2succes) [inline schema/me/conversation.js]: Can a message have multiple attachments?

File: schema/me/conversation.js (hunk 5 above)
Comment (mzikherman) [inline schema/me/conversation.js]: Yea. This is likely to change soon, so at some point we'd remove this (but we'd keep the `is_from_user` still resolving properly, so it should be transparent to Emission).

File: schema/me/conversation.js (hunk 6 above)
Comment (mzikherman) [inline schema/me/conversation.js]: Yup!

**Comment by mzikherman** [comment]
> codewise LGTM! we'll have to implement it in emission and see if the new impulse endpoint works as intended

Yea I have the naive changes on Emission working, where there isn't an added Relay fragment (basically it just replaces `snippet` with `raw_text`)...and it works!

I'm sure there'll be some weird gotchas/Relay stuff happening though.

**Comment by mzikherman** [comment]
Alright @l2succes I'll add some specs (!) but this is working with impulse on staging.

cc @sarahscott now we have:

screen shot 2017-06-23 at 4 22 05 pm

(`is_from_user` is working).

I pushed this branch up to `artsy/metaphysics` as well in case anyone wants/needs to use it locally (with the `sync-schema`) thing.

**Comment by sarahscott** [comment]
@mzikherman is this ready to merge?

File: package.json [hunk 7]
```diff
@@ -18,7 +18,7 @@
     "lint": "eslint .",
     "flow": "flow check",
     "danger": "danger",
-    "fix": "eslint . --fix",
+    "fix": "yarn eslint . --fix",
```
Comment (l2succes) [inline package.json]: do we need `yarn` here or you were experimenting?

Comment (mzikherman) [inline package.json]: Ah, experimenting. Turns out all my failures where from files in the `build/` directory, so i added it to the ignore. Will undo.

Comment (l2succes) [inline package.json]: Ok cool, otherwise everything else looks good so let's merge after removing that

**Comment by mzikherman** [comment]
Ok, so this should be mergable, but it'll break people's Emission/inbox that's pointing to staging, since the schema is updated.

To be a good citizen, should we first get https://github.com/artsy/emission/pull/630 merged (or at least 🍏 ).

Then, it should be relatively easy for people to update, and not have any intermediate states. I'll take a look at the Emission PR now to see what's up.

**Comment by mzikherman** [comment]
k merging!

**Pull request merged**